server-spi/src/main/java/org/keycloak/theme/Theme.java

```diff
@@ -63,6 +64,19 @@ enum Type { LOGIN, ACCOUNT, ADMIN, EMAIL, WELCOME, COMMON };
      */
     Properties getMessages(String baseBundlename, Locale locale) throws IOException;
 
+    /**
+     * Retrieve localized messages from a message bundle named "messages".
+     * <p>
+     * Returns messages for all relevant locales grouped by locale. This is different from
+     * {@link #getMessages(Locale)}, which returns already merged messages.
+     * </p>
+     * 
+     * @param locale The locale of the desired message bundle.
+     * @return The localized messages from the bundle, grouped by locale.
+     * @throws IOException If bundle can not be read.
+     */
+    Map<Locale, Properties> getGroupedMessages(Locale locale) throws IOException;
```

Comment:
How about to instead add to the `Theme` interface some method like `Properties getEnhancedProperties(RealmModel realm, Locale locale)` ?

This can be fine as it looks that all the usages of the method `getGroupedMessages` in this PR are doing something like:
```
Map<Locale, Properties> themeMessages = theme.getGroupedMessages(locale);
Properties messages = LocaleUtil.enhancePropertiesWithRealmLocalizationTexts(realm, locale, themeMessages);
```
which can be simplified IMO just to something like:
```
Properties messages = theme.getEnhancedMessages(realm, locale);
```

In this case, the method can be also declared as `default` on the theme interface with just calling `getProperties(locale)` as that is sufficient for both of `ClassLoaderTheme` and `FolderTheme` IMO and just `ExtendedTheme` will need to implement it in a same way like you're doing?

WDYT?

Reply:
Thanks for the hint.

Introduced a method Properties `getEnhancedMessages(RealmModel realm, Locale locale)` on the Theme interface and used it instead of `getGroupedMessage`. Makes the usage much simpler.

However, I did not introduce a default implementation, because otherwise the Theme module (keycloak-server-spi) would need a dependency on keycloak-services (due to LocaleUtil). And that dependency would be circular.

Reply:
Thanks!